날씨앱 [STEP2] 요한

dpwns1234 opened the pull request:
# 날씨 앱 [STEP2] 요한
리뷰어: @coryKim0829

안녕하세요 코리!
STEP2 PR 보냅니다. 잘 부탁드립니다!


## 파일구조📁
- Secrets.xcconfig
- Delegate
    - LocationDelegate
- Util
    - LocationDataManager
- Network
    - WeatherApiClient
    - WeatherNetworkManager
    - NetworkError
- Model
    - WeatherType
    - WeatherToday
    - WeatherForecast
    - WeatherCommonInformation
- Controller
    - WeatherViewController


## 구현내용
- [x] `CoreLocation`을 통해 현재 좌표 가져오기
- [x] `CoreLocation`을 통해 주소 가져오기
- [x] `URLSession`을 통해 API 호출 (HTTP GET 메서드 명시적 호출)
- [x] 전달받은 `JSON` 데이터 파싱 및 모델에 매칭

### Delegate 패턴을 사용하여 LocationDataManager와 WeatherNetworkManager를 통신하도록 구현

```swift 
// LocationDataManagerDelegate.swfit

protocol LocationDataManagerDelegate: AnyObject {
    func location(_ manager: LocationDataManager, didLoad coordinate: CLLocationCoordinate2D)
}
```


```swift
// LocationDataManager.swift 

final class LocationDataManager : NSObject {
    weak var locationDelegate: LocationDataManagerDelegate?
    ...
}
```

```swift 
// WeatherViewController.swift

final class WeatherViewController: UIViewController {
    private let networkManager = WeatherNetworkManager()
    private let locationDataManager = LocationDataManager()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        locationDataManager.locationDelegate = self
    }
}

extension WeatherViewController: LocationDataManagerDelegate {
    func location(_ manager: LocationDataManager, didLoad coordinate: CLLocationCoordinate2D) {
        networkManager.loadWeatherData(type: WeatherType.weatherToday, coord: coordinate)
    }
}
```


## 고민한 점
LocationDataManager의 lookUpCurrentAddress 메서드의 depth가 깊다고 생각되어 함수로 분리하려 하였습니다. 그래서 reverseGeocodeLocation()의 후행클로저를 함수로 분리하려 하였지만, lookUpCurrentAddress()의 인자인 completionHandler를 같이 넘겨줄 수 없어서 현재의 코드를 유지하였습니다.
```swift 
private func lookUpCurrentAddress(completionHandler: @escaping (CLPlacemark?) -> Void ) {
    if let lastLocation = locationManager.location {
        let geocoder = CLGeocoder()

        geocoder.reverseGeocodeLocation(lastLocation) { (placemarks, error) in
            if error == nil {
                guard let placemarks else { return }
                let firstLocation = placemarks[0]
                completionHandler(firstLocation)
            } else {
                completionHandler(nil)
            }
        }
    } else {
        completionHandler(nil)
    }
}
```

dpwns1234 committed:
refactor: try force unwrapping -> do-catch문으로 예외처리

dpwns1234 committed:
Merge pull request #1 from dpwns1234/STEP1

Step1

dpwns1234 committed:
fix: secrets 파일 수정

dpwns1234 committed:
Merge branch 'main' into STEP1

dpwns1234 committed:
fix: secrets 파일 추가

dpwns1234 committed:
Merge branch 'STEP2' of https://github.com/dpwns1234/ios-weather-forecast into STEP2

dpwns1234 committed:
refactor: 구체적인 객체명으로 변경

dpwns1234 committed:
feat: 위치 정보를 가져오는 LocationDataManager 클래스 생성

dpwns1234 committed:
feat: LocationDataManager와 ViewController와 연결하는 LocationDelegate 구현

- LocationDataManager에서 불러낸 위도와 경도를 VC에 전달

dpwns1234 committed:
refactor: delegate 패턴에 맞게 인자 객체 수정

dpwns1234 committed:
fix: 데이터 모델 수정

- sea_level, grand_level 필드가 없을 수도 있었음. -> 옵셔널 타입으로 변경

dpwns1234 committed:
feat: error 타입 enum 구현

dpwns1234 committed:
refactor: HTTP GET requset로 api 통신 요청

dpwns1234 committed:
feat: CoreLocation을 통해 주소정보 가져오는 기능 구현

dpwns1234 committed:
refactor: handler 함수로 분리

- 가독성 증가

dpwns1234 committed:
refactor: lookUpCurrentLocation() 메서드 네이밍 변경 및 placemarks 오타 수정

dpwns1234 committed:
refactor: Delegate 네이밍 변경

dpwns1234 committed:
refactor: 함수 리팩터

- 함수명 변경
- 함수 인자 순서 변경
- private 선언

dpwns1234 committed:
fix: 모델 옵셔널타입 수정

dpwns1234 committed:
refactor: 위치 주소 프로퍼티의 옵셔널 풀어주도록 구현

dpwns1234 committed:
chore: TODO 주석 지움

corykim0829 commented on WeatherForecast/WeatherForecast/Util/LocationDataManager.swift:
```diff
@@ -0,0 +1,101 @@
+    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
+        if let coordinate = locations.last?.coordinate {
+            locationDelegate?.location(self, didLoad: coordinate)
+            lookUpCurrentAddress(completionHandler: viewCurrentAddress)
```
lookUpCurrentAddress 함수 인자로 viewCurrentAddress함수를 받는 부분이 조금 어색하네요!

dpwns1234 replied:
currentAddress을 찾았을 때(= complete), viewCurrentAddress() 함수를 실행하라는 의미로 위와 같이 작성하였었습니다. 현재는 viewCurrentAddress() 를 ViewController로 이동하긴 했는데, 코리의 의도대로 수정했는지는 잘 모르겠습니다.
혹시 어떤 부분이 어색한지, 어떤 방향으로 수정해야하는지 알 수 있을까요?

corykim0829 replied:
이번에 수정하신 lookUpCurrentAddress completion 파라미터를 클로저로 처리하신게 훨씬 깔끔하고 좋습니다!
잘 수정하셨습니다! 👍

corykim0829 commented on WeatherForecast/WeatherForecast/Util/LocationDataManager.swift:
```diff
@@ -0,0 +1,101 @@
+                let subThoroughfare = placemark.subThoroughfare
+            {
+                let address = "\(country) \(administrativeArea) \(locality) \(subLocality) \(thoroughfare) \(subThoroughfare)"
+                print(address)
```
데이터를 확인하는 것도 좋지만, 해당 데이터를 컨트롤러에게 어떻게 전달할지도 고민해보면 좋을 것 같습니다!

dpwns1234 replied:
locationDelegate를 사용해서 데이터를 전달하고 address를 출력하는 해당 메서드를 viewController에서 실행하도록 수정하였습니다!

corykim0829 commented on WeatherForecast/WeatherForecast/Util/LocationDataManager.swift:
```diff
@@ -0,0 +1,101 @@
+            
+            geocoder.reverseGeocodeLocation(lastLocation) { (placemarks, error) in
+                if error == nil {
+                    guard let placemarks else { return }
```
return 되는 경우의 예외 처리도 필요해보입니다

dpwns1234 replied:
completeHandler에 nil을 전달함으로써 VC에 있는 곳에서 에러처리 해주었습니다!

corykim0829 commented on WeatherForecast/WeatherForecast/Util/LocationDataManager.swift:
```diff
@@ -0,0 +1,101 @@
+// MARK: - Get Address
+
+extension LocationDataManager {
+    private func lookUpCurrentAddress(completionHandler: @escaping (CLPlacemark?) -> Void ) {
```
이 함수 안에 if let이 많이 사용되는게 보이는데, guard let을 활용하면 조금 더 깔끔하게 정리할 수 있을 것 같아요! 이건 참고만 하세요!

corykim0829 commented on WeatherForecast/WeatherForecast/Util/LocationDataManager.swift:
```diff
@@ -0,0 +1,101 @@
+                print(address)
+            }
+        } else {
+            print("location 가져오기 실패")
```
이 부분도 예외 처리를 어떻게 컨트롤러에 전달할지도 고민해보면 좋을 것 같아요

dpwns1234 replied:
해당 메서드를 ViewController에 옮기고, delegate로 전달하였습니다!

corykim0829 commented on WeatherForecast/WeatherForecast/Util/LocationDataManager.swift:
```diff
@@ -0,0 +1,101 @@
+            switch manager.authorizationStatus {
+            case .authorizedWhenInUse:
+                manager.startUpdatingLocation()
+                break
```
생략 가능한 break는 지워주세요

corykim0829 commented on WeatherForecast/WeatherForecast/Util/LocationDataManager.swift:
```diff
@@ -0,0 +1,101 @@
+            case .authorizedWhenInUse:
+                manager.startUpdatingLocation()
+                break
+            case .restricted, .denied:
+                // showRequestLocationServiceAlert()
+                // disableLocationFeatures()
```
예외 처리 고민해주시면 좋을 것 같아요!

dpwns1234 replied:
위치 권한을 거부했을 때는, 설정창으로 이동하도록 하여 권한을 허락하는 alert을 띄워주도록 수정하였습니다!

corykim0829 commented on WeatherForecast/WeatherForecast/Util/LocationDataManager.swift:
```diff
@@ -0,0 +1,101 @@
+    
+    override init() {
+        super.init()
+        // 절전 기능 활성화
+        locationManager.distanceFilter = CLLocationDistanceMax
+        locationManager.allowsBackgroundLocationUpdates = false
```
주석도 좋지만, 함수로 분리해서 더 깔끔하게 정리하면 좋을 것 같습니다!

dpwns1234 replied:
감사합니다! 함수로 분리하였습니다.

corykim0829 commented on WeatherForecast/WeatherForecast/Network/WeatherApiClient.swift:
```diff
@@ -5,21 +5,28 @@
     }
+    
+    static func makeRequest(weatherType: WeatherType, coord: CLLocationCoordinate2D) throws -> URLRequest {
+        guard let url = makeURL(coord: coord, weatherType: weatherType) else { throw NetworkError.invailedURL }
```
guard-else 뒤 대괄호 부분은 다음과 같이 줄바꿈을 해주세요 명확하게 예외 처리가 되는 부분을 줄바꿈으로 처리하면 가독성이 더 좋습니다
```Swift
else {
    ...
    return
}
```

corykim0829 commented on WeatherForecast/WeatherForecast/Network/NetworkError.swift:
```diff
@@ -0,0 +1,19 @@
+
+import Foundation
+
+enum NetworkError: Error, CustomStringConvertible {
+    case invailedURL
```
invalidURL 외에도 다른 에러 상황은 없을까요?

dpwns1234 replied:
한 가지 궁금한 것이 있습니다! 저는 발생할 수 있는 에러들을 ViewController로 보내어 처리하려고 하였으나 throw를 날리기 어려운 상황이나 여러 어려움으로 한 곳이 아닌 여러 곳에서 do-catch를 해야하는 상황입니다. 
제 생각에는 발생할 수 있는 에러들을 throw 날리고 do-catch 구문을 Controller(ViewController)에서 혹은 한 곳에서 따로 처리를 하는 것이 깔끔하다고 생각하는데, 굳이 신경 쓸 필요 없는 것인지, 코리는 어떻게 생각하는지 궁금합니다!

corykim0829 replied:
지금 NetworkError가 발생하는 곳은 WeatherNetworkManager 밖에 없는 상황에서, WeatherNetworkManager의 loadWeatherData 함수를 호출할 때 발생할 수 있는 에러 상황에 대한 예외 처리를 loadWeatherData 이 함수 내에서 처리하고 이에 대한 결과값만 ViewController로 넘겨주면 좋을 것 같습니다!
ViewController는 데이터 또는 에러만 받아서 처리해주면 되니까요!

dpwns1234 replied:
아하 이해했습니다! 에러에 대한 예외 처리는 그 곳에서, 결과 혹은 보여줄 결과(alert 등)는 뷰컨에서 처리하도록 수정하겠습니다!

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/WeatherViewController.swift:
```diff
@@ -5,13 +5,22 @@
+
+// MARK: - LocationDataManagerDelegate
+
+extension WeatherViewController: LocationDataManagerDelegate {
+    func location(_ manager: LocationDataManager, didLoad coordinate: CLLocationCoordinate2D) {
```
delegate로 데이터 전달 구현을 잘하셨습니다! 👍

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/WeatherViewController.swift:
```diff
@@ -5,13 +5,22 @@
+
+extension WeatherViewController: LocationDataManagerDelegate {
+    func location(_ manager: LocationDataManager, didLoad coordinate: CLLocationCoordinate2D) {
+        networkManager.loadWeatherData(type: WeatherType.weatherToday, coord: coordinate)
```
networkManager를 통해서 받은 데이터는 어떻게 전달할건지도 고민해보면 좋을 것 같아요

dpwns1234 replied:
delegate 패턴으로 데이터를 뷰컨에 전달해주도록 수정하였습니다!

corykim0829 reviewed (changes requested):
수고 많으셨습니다!

Manager 객체들로 분리해서 일을 처리하는 것은 좋으나 마무리까지 모두 Manager에서 처리되고 있어서 아쉽습니다!
아직 UI 구현은 아니지만, 그래도 데이터들을 컨트롤러에 전달해서 컨트롤러에서 print로 확인하면 좋을 것 같습니다!

delegate 사용도 잘 하시고, 조금만 더 정리하면 더 가독성 좋고 깔끔한 코드가 될 것 같습니다!

dpwns1234 committed:
refactor: LocationDataManager에서 주소를 출력하는 메서드를 viewController로 이동

- LocationDelegate를 사용하여 구현

dpwns1234 committed:
refactor: lookUpCurrentAddress error처리

- current loaction을 찾는 메서드에서 주소를 찾지 못할 경우 handler의 인자로 nil을 전달하고, 인자(placemark)가 nil이라면 에러 처리를 해준다.

dpwns1234 committed:
refactor: 절전 기능 함수로 분리

dpwns1234 committed:
refactor: 위치 권한 거부에 따른 alert 띄워주기

dpwns1234 committed:
refactor: weatherNetworkManager에서 load한 데이터를 viewController로 전달

- delegate 패턴으로 데이터를 전달함.
- ViewController에서 data를 print

dpwns1234 committed:
refactor: Location group 생성 및 delegate 파일 각각 이동

dpwns1234 committed:
refactor: LocationDelegate의 함수를 delegate 함수 형식에 맞게 refactor

dpwns1234 committed:
refactor: Error의 group 이름 변경

dpwns1234 committed:
refactor: 수정된 delegate 함수 네이밍 VC에 적용

- viewCurrentAddress -> location()

corykim0829 commented on WeatherForecast/WeatherForecast/Location/LocationDataManagerDelegate.swift:
```diff
@@ -0,0 +1,14 @@
+import CoreLocation
+
+protocol LocationDataManagerDelegate: AnyObject {
+    func location(_ manager: LocationDataManager, didLoad coordinate: CLLocationCoordinate2D)
+    func loaction(_ manager: LocationDataManager, didComplete placemark: CLPlacemark?)
+    func viewRequestLocationSettingAlert()
```
delegate를 잘 활용해서 ViewController와 데이터 전달을 잘 하고 계십니다! 👍
조금 아쉬운 점은 delegate의 함수명만 보고 조금 더 기능을 유추할 수 있도록 개선해보면 더 좋은 코드가 될 것 같습니다!

현재 Swift에서 주로 delegate 함수들의 네이밍을 어떻게 하는지 참고해서 수정해보면 더 좋을 것 같습니다!
현재 didLoad~ didComplete~ 만 보면 목적어가 정확하게 나타나지 않아서 가독성이 조금 아쉬운 것 같아요!

dpwns1234 replied:
didLoadCoordinate, didCompletePlacmark 로 수정하였습니다! 감사합니다 😊

corykim0829 reviewed (approved):
수고 많으셨습니다!!

추가로 남겨드린 커멘트는 다음 스텝 진행하시면서 수정해주세요!!

항상 예외 처리를 고민하시는 습관을 가지시면 좋을 것 같습니다!

고생 많으셨어요!
